### ISSUE TITLE
mat: failures on go master with triangular storage mismatch panic

### ISSUE THREAD

**Issue by kortschak** [gonum/gonum#522]
In the tests for #519, I see failures on go master (go version devel +c08b01ecb4 Wed Jun 6 20:35:23 2018 +0000 linux/amd64 - previous build for us passed: go version devel +7c1f361e2, so the bisection should not be too hard). In all cases, I don't understand how this can be happening. I suspect it may be a compiler bug.

```
panic: matrix: triangular storage mismatch [recovered]
	panic: matrix: triangular storage mismatch

goroutine 5 [running]:
testing.tRunner.func1(0xc0000b8b00)
	/home/travis/.gimme/versions/go/src/testing/testing.go:792 +0x387
panic(0x6902e0, 0xc0001b4110)
	/home/travis/.gimme/versions/go/src/runtime/panic.go:494 +0x1b9
gonum.org/v1/gonum/mat.(*TriDense).reuseAs(0xc0000ca780, 0x3, 0xc0000a6600)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/triangular.go:263 +0x1dc
gonum.org/v1/gonum/mat.(*TriDense).ScaleTri(0xc0000ca780, 0x3fe6a09e667f3bcd, 0x70f620, 0xc0000ca740)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/triangular.go:427 +0x70
gonum.org/v1/gonum/mat.(*Cholesky).Scale(0xc0000a5ef8, 0x3fe0000000000000, 0xc0000a5f18)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/cholesky.go:344 +0xb4
gonum.org/v1/gonum/mat.TestCholeskyScale(0xc0000b8b00)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/cholesky_test.go:532 +0x1e5
testing.tRunner(0xc0000b8b00, 0x6ccdc0)
	/home/travis/.gimme/versions/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/home/travis/.gimme/versions/go/src/testing/testing.go:878 +0x353
FAIL	gonum.org/v1/gonum/mat	0.152s
```

```
=== RUN   TestCmaEsChol
panic: matrix: triangular storage mismatch

goroutine 6 [running]:
gonum.org/v1/gonum/mat.(*TriDense).reuseAs(0xc00005c900, 0xa, 0x0)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/triangular.go:263 +0x1dc
gonum.org/v1/gonum/mat.(*TriDense).ScaleTri(0xc00005c900, 0x3fef6d8944f11369, 0x65ca20, 0xc00005c900)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/triangular.go:427 +0x70
gonum.org/v1/gonum/mat.(*Cholesky).Scale(0xc000001620, 0x3feeddb0e70596d5, 0xc000001620)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/cholesky.go:344 +0xb4
gonum.org/v1/gonum/optimize.(*CmaEsChol).update(0xc000001500, 0x9, 0x10)
	/home/travis/gopath/src/gonum.org/v1/gonum/optimize/cmaes.go:443 +0x836
gonum.org/v1/gonum/optimize.(*CmaEsChol).RunGlobal(0xc000001500, 0xc00005e0c0, 0xc00005e120, 0xc00000a2e0, 0x1, 0x1)
	/home/travis/gopath/src/gonum.org/v1/gonum/optimize/cmaes.go:344 +0x19d
gonum.org/v1/gonum/optimize.minimizeGlobal.func1(0x1, 0xc00005c840, 0x0, 0xa, 0x65c460, 0xc000001500, 0xc00005e0c0, 0xc00005e120)
	/home/travis/gopath/src/gonum.org/v1/gonum/optimize/global.go:189 +0x179
created by gonum.org/v1/gonum/optimize.minimizeGlobal
	/home/travis/gopath/src/gonum.org/v1/gonum/optimize/global.go:182 +0x15e
FAIL	gonum.org/v1/gonum/optimize	0.005s
```

```
--- FAIL: TestWishart (0.00s)
panic: matrix: triangular storage mismatch [recovered]
	panic: matrix: triangular storage mismatch

goroutine 5 [running]:
testing.tRunner.func1(0xc00008c100)
	/home/travis/.gimme/versions/go/src/testing/testing.go:792 +0x387
panic(0x5c6e40, 0xc00002a4e0)
	/home/travis/.gimme/versions/go/src/runtime/panic.go:494 +0x1b9
gonum.org/v1/gonum/mat.(*TriDense).reuseAs(0xc00005cb80, 0x2, 0xc00005cb00)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/triangular.go:263 +0x1dc
gonum.org/v1/gonum/mat.(*TriDense).MulTri(0xc00005cb80, 0x61b4e0, 0xc00005cb80, 0x61b4e0, 0xc00008e158)
	/home/travis/gopath/src/gonum.org/v1/gonum/mat/triangular.go:388 +0x1b4
gonum.org/v1/gonum/stat/distmat.(*Wishart).RandChol(0xc00008e120, 0x0, 0x3d06849b86a12b9b)
	/home/travis/gopath/src/gonum.org/v1/gonum/stat/distmat/wishart.go:196 +0x267
gonum.org/v1/gonum/stat/distmat.TestWishart(0xc00008c100)
	/home/travis/gopath/src/gonum.org/v1/gonum/stat/distmat/wishart_test.go:73 +0xb5f
testing.tRunner(0xc00008c100, 0x5f4c18)
	/home/travis/.gimme/versions/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/home/travis/.gimme/versions/go/src/testing/testing.go:878 +0x353
FAIL	gonum.org/v1/gonum/stat/distmat	0.016s
```

Also, this, which may or may not be related (the failures above do involve Cholesky, but they are panics):

```
=== RUN   TestCholeskyInverseTo
--- FAIL: TestCholeskyInverseTo (0.00s)
    cholesky_test.go:304: SymDense times Cholesky inverse not identity. Norm diff = 2.4565376112725565
    cholesky_test.go:304: SymDense times Cholesky inverse not identity. Norm diff = 3.0519897080345144
    cholesky_test.go:304: SymDense times Cholesky inverse not identity. Norm diff = 8.120946693591646
```

**Comment by kortschak** [comment]
Bisected to golang/go@c08b01ecb4488fb3a95fd5cc7baa8b31812e7b76. I'll try to look into this later today, but may not get to it until tomorrow.

/cc @dr2chase

```
commit c08b01ecb4488fb3a95fd5cc7baa8b31812e7b76
Author: David Chase <drchase@google.com>
Date:   Wed Jun 6 12:38:35 2018 -0400

    cmd/compile: fix panic-okay-to-inline change; adjust tests
    
    This line of the inlining tuning experiment
    https://go-review.googlesource.com/c/go/+/109918/1/src/cmd/compile/internal/gc/inl.go#347
    was incorrectly rewritten in a later patch to use the call
    cost, not the panic cost, and thus the inlining of panic
    didn't occur when it should.  I discovered this when I
    realized that tests should have failed, but didn't.
    
    Fix is to make the correct change, and also to modify the
    tests that this causes to fail.  One test now asserts the
    new normal, the other calls "ppanic" instead which is
    designed to behave like panic but not be inlined.
    
    Change-Id: I423bb7f08bd66a70d999826dd9b87027abf34cdf
    Reviewed-on: https://go-review.googlesource.com/116656
    Run-TryBot: David Chase <drchase@google.com>
    TryBot-Result: Gobot Gobot <gobot@golang.org>
    Reviewed-by: Keith Randall <khr@golang.org>
```

**Comment by kortschak** [comment]
Minimal gonum reproducer:

```
package main

import "gonum.org/v1/gonum/mat"

func main() {
	ta := mat.NewTriDense(3, mat.Upper, nil)
	tb := mat.NewTriDense(3, mat.Upper, nil)

	ta.ScaleTri(1, tb)
}
```

**Comment by kortschak** [comment]
Fixed by golang/go@c359d759a77fcec1457f3eb5c5d04fb74f47dad4.
